File: src/services/ads.service.ts
import { v4 as uuidv4 } from 'uuid';
import { query, queryOne, execute } from '../config/db';
import { fileService } from './file.service';
import { isBrandAllowed, isCityAllowed, isConditionAllowed, normaliseBrand, normaliseCity } from '../utils/constants';

// ─── Types ────────────────────────────────────────────────────────────────────

interface ListQuery {
  page?:      number;
  limit?:     number;
  q?:         string;
  brand?:     string;
  condition?: string;
  city?:      string;
  min_price?: number;
  max_price?: number;
  sort?:      'newest' | 'oldest' | 'price_asc' | 'price_desc';
}

interface CreateAdInput {
  title:          string;
  brand:          string;
  model:          string;
  price:          number;
  condition:      string;
  city:           string;
  description:    string;
  contact_number: string;
  photoFiles:     Express.Multer.File[];  // uploaded via multipart
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

function validateAdFields(data: Partial<CreateAdInput & { condition?: string }>, requireAll = true) {
  if (data.brand !== undefined && !isBrandAllowed(data.brand))
    throw Object.assign(new Error('Brand not in allowed list.'), { status: 422, code: 'invalid_brand' });
  if (data.city !== undefined && !isCityAllowed(data.city))
    throw Object.assign(new Error('City not in allowed list.'), { status: 422, code: 'invalid_city' });
  if (data.condition !== undefined && !isConditionAllowed(data.condition))
    throw Object.assign(new Error('Invalid condition value.'), { status: 422, code: 'invalid_condition' });
  if (requireAll && (!data.photoFiles || data.photoFiles.length === 0))
    throw Object.assign(new Error('At least 1 photo is required.'), { status: 422, code: 'validation_error' });
  if (data.photoFiles && data.photoFiles.length > 5)
    throw Object.assign(new Error('Maximum 5 photos allowed.'), { status: 422, code: 'too_many_photos' });
}

const SORT_MAP: Record<string, string> = {
  newest:     'created_at DESC',
  oldest:     'created_at ASC',
  price_asc:  'price ASC',
  price_desc: 'price DESC',
};

/** Resolve stored photo paths to serve URLs */
function resolvePhotos(row: any, baseUrl: string): string[] {
  const paths: string[] = typeof row.photo_urls === 'string'
    ? JSON.parse(row.photo_urls)
    : (row.photo_urls ?? []);
  return paths.map(p => fileService.serveUrl(p, baseUrl));
}

function formatAd(row: any, baseUrl = '') {
  return {
    id:             row.id,
    title:          row.title,
    brand:          row.brand,
    model:          row.model,
    price:          parseFloat(row.price),
    condition:      row.condition_val,
    city:           row.city,
    description:    row.description,
    contact_number: row.contact_number,
    photo_urls:     resolvePhotos(row, baseUrl),
    user_id:        row.user_id,
    seller_name:    row.seller_name,
    is_active:      Boolean(row.is_active),
    created_at:     row.created_at instanceof Date ? row.created_at.toISOString() : row.created_at,
    updated_at:     row.updated_at instanceof Date ? row.updated_at.toISOString() : row.updated_at,
  };
}

// ─── Service ──────────────────────────────────────────────────────────────────

export const adsService = {
  // ── Browse / list ─────────────────────────────────────────────────────────
  async list(params: ListQuery, baseUrl = '') {
    const page  = Math.max(1, params.page  || 1);
    const limit = Math.min(50, Math.max(1, params.limit || 20));
    const offset = (page - 1) * limit;
    const sort  = SORT_MAP[params.sort || 'newest'] || SORT_MAP.newest;

    const conditions: string[] = ['is_active = 1'];
    const values: any[]        = [];

    if (params.q) {
      conditions.push('MATCH(title, brand, model, city) AGAINST(? IN BOOLEAN MODE)');
      values.push(`${params.q}*`);
    }
    if (params.brand)     { conditions.push('LOWER(brand) = LOWER(?)');     values.push(params.brand); }
    if (params.condition) { conditions.push('condition_val = ?');            values.push(params.condition); }
    if (params.city)      { conditions.push('LOWER(city) = LOWER(?)');      values.push(params.city); }
    if (params.min_price !== undefined) { conditions.push('price >= ?');     values.push(params.min_price); }
    if (params.max_price !== undefined) { conditions.push('price <= ?');     values.push(params.max_price); }

    const where = `WHERE ${conditions.join(' AND ')}`;

    const countRows = await query<{ total: number }>(
      `SELECT COUNT(*) AS total FROM phone_ads ${where}`, values
    );
    const total       = countRows[0]?.total || 0;
    const total_pages = Math.ceil(total / limit);

    const rows = await query(
      `SELECT * FROM phone_ads ${where} ORDER BY ${sort} LIMIT ? OFFSET ?`,
      [...values, limit, offset]
    );

    return {
      data: rows.map(r => formatAd(r, baseUrl)),
      meta: { page, limit, total, total_pages, has_more: page < total_pages },
    };
  },

  // ── Single ad ─────────────────────────────────────────────────────────────
  async getById(id: string, baseUrl = '') {
    const row = await queryOne<any>(
      'SELECT * FROM phone_ads WHERE id = ? AND is_active = 1', [id]
    );
    if (!row) throw Object.assign(new Error('Ad not found.'), { status: 404, code: 'not_found' });
    return formatAd(row, baseUrl);
  },

  // ── Create ────────────────────────────────────────────────────────────────
  async create(userId: string, sellerName: string, data: CreateAdInput, baseUrl = '') {
    validateAdFields(data, true);

    // Daily limit: max 20 ads per user per day
    const todayCount = await queryOne<{ cnt: number }>(
      `SELECT COUNT(*) AS cnt FROM phone_ads WHERE user_id = ? AND DATE(created_at) = CURDATE()`,
      [userId]
    );
    if ((todayCount?.cnt || 0) >= 20) {
      throw Object.assign(new Error('Daily limit of 20 ads reached.'), { status: 429, code: 'rate_limit_exceeded' });
    }

    // Upload each photo to files table
    const photoPaths: string[] = [];
    for (const file of data.photoFiles) {
      const saved = await fileService.save({
        originalName: file.originalname,
        mimeType:     file.mimetype,
        size:         file.size,
        buffer:       file.buffer,
        module:       'ads',
        uploadedBy:   userId,
      });
      photoPaths.push(saved.path);
    }

    const id = uuidv4();
    const now = new Date();
    await execute(
      `INSERT INTO phone_ads
         (id, title, brand, model, price, condition_val, city, description,
          contact_number, photo_urls, user_id, seller_name, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        id,
        data.title.trim(),
        normaliseBrand(data.brand),
        data.model.trim(),
        data.price,
        data.condition,
        normaliseCity(data.city),
        data.description.trim(),
        data.contact_number,
        JSON.stringify(photoPaths),   // store paths, not URLs
        userId,
        sellerName,
        now,
        now
      ]
    );

    const row = await queryOne<any>('SELECT * FROM phone_ads WHERE id = ?', [id]);
    return formatAd(row, baseUrl);
  },

  // ── Update ────────────────────────────────────────────────────────────────
  async update(
    id: string,
    userId: string,
    data: Partial<Omit<CreateAdInput, 'photoFiles'>> & { newPhotoFiles?: Express.Multer.File[] },
    baseUrl = ''
  ) {
    const existing = await queryOne<any>('SELECT * FROM phone_ads WHERE id = ?', [id]);
    if (!existing) throw Object.assign(new Error('Ad not found.'), { status: 404, code: 'not_found' });
    if (existing.user_id !== userId) throw Object.assign(new Error('You are not allowed to edit this ad.'), { status: 403, code: 'forbidden' });

    if (data.brand)     validateAdFields({ brand: data.brand }, false);
    if (data.city)      validateAdFields({ city: data.city }, false);
    if (data.condition) validateAdFields({ condition: data.condition }, false);

    const fields: string[] = [];
    const values: any[]    = [];

    if (data.title !== undefined)          { fields.push('title = ?');          values.push(data.title.trim()); }
    if (data.brand !== undefined)          { fields.push('brand = ?');          values.push(normaliseBrand(data.brand)); }
    if (data.model !== undefined)          { fields.push('model = ?');          values.push(data.model.trim()); }
    if (data.price !== undefined)          { fields.push('price = ?');          values.push(data.price); }
    if (data.condition !== undefined)      { fields.push('condition_val = ?');  values.push(data.condition); }
    if (data.city !== undefined)           { fields.push('city = ?');           values.push(normaliseCity(data.city)); }
    if (data.description !== undefined)    { fields.push('description = ?');    values.push(data.description.trim()); }
    if (data.contact_number !== undefined) { fields.push('contact_number = ?'); values.push(data.contact_number); }

    // Replace photos if new ones uploaded
    if (data.newPhotoFiles && data.newPhotoFiles.length > 0) {
      if (data.newPhotoFiles.length > 5)
        throw Object.assign(new Error('Maximum 5 photos allowed.'), { status: 422, code: 'too_many_photos' });

      // Delete old blobs
      const oldPaths: string[] = typeof existing.photo_urls === 'string'
        ? JSON.parse(existing.photo_urls) : (existing.photo_urls ?? []);
      for (const p of oldPaths) {
        await fileService.deleteByPath(p).catch(() => { });
      }

      // Save new blobs
      const newPaths: string[] = [];
      for (const file of data.newPhotoFiles) {
        const saved = await fileService.save({
          originalName: file.originalname,
          mimeType:     file.mimetype,
          size:         file.size,
          buffer:       file.buffer,
          module:       'ads',
          uploadedBy:   userId,
        });
        newPaths.push(saved.path);
      }
      fields.push('photo_urls = ?');
      values.push(JSON.stringify(newPaths));
    }

    if (fields.length > 0) {
      fields.push('updated_at = ?');
      values.push(new Date());
      values.push(id);
      await execute(`UPDATE phone_ads SET ${fields.join(', ')} WHERE id = ?`, values);
    }

    const updated = await queryOne<any>('SELECT * FROM phone_ads WHERE id = ?', [id]);
    return formatAd(updated, baseUrl);
  },

  // ── Soft delete ───────────────────────────────────────────────────────────
  async softDelete(id: string, userId: string) {
    const existing = await queryOne<any>('SELECT id, user_id FROM phone_ads WHERE id = ?', [id]);
    if (!existing) throw Object.assign(new Error('Ad not found.'), { status: 404, code: 'not_found' });
    if (existing.user_id !== userId) throw Object.assign(new Error('You are not allowed to delete this ad.'), { status: 403, code: 'forbidden' });
    await execute('UPDATE phone_ads SET is_active = 0 WHERE id = ?', [id]);
  },

  // ── My ads ────────────────────────────────────────────────────────────────
  async myAds(userId: string, baseUrl = '') {
    const rows = await query<any>(
      'SELECT * FROM phone_ads WHERE user_id = ? ORDER BY created_at DESC', [userId]
    );
    return {
      data: rows.map(r => formatAd(r, baseUrl)),
      meta: { total: rows.length },
    };
  },
};

File: EasyPhones.Application/Services/AdsService.cs
using System.Text.Json;
using EasyPhones.Application.Constants;
using EasyPhones.Application.Dtos.AdsDtos;
using EasyPhones.Application.Exceptions;
using EasyPhones.Application.Wrappers;
using EasyPhones.Domain.Entities;
using EasyPhones.Domain.Enums;
using EasyPhones.Domain.Repositories;

namespace EasyPhones.Application.Services;

public class AdsService
{
    private readonly IAdRepository _ads;
    private readonly IFileService _files;

    public AdsService(IAdRepository ads, IFileService files)
    {
        _ads = ads;
        _files = files;
    }

    // ── List ──────────────────────────────────────────────────────────────────
    public async Task<PagedResponse<AdDto>> ListAsync(AdListRequestDto req, string baseUrl, CancellationToken ct)
    {
        var page  = Math.Max(1, req.Page);
        var limit = Math.Clamp(req.Limit, 1, 50);

        var query = new AdListQuery(page, limit, req.Q, req.Brand, req.Condition, req.City,
            req.MinPrice, req.MaxPrice, req.Sort ?? "newest");

        var (items, total) = await _ads.ListAsync(query, ct);
        var totalPages = (int)Math.Ceiling((double)total / limit);

        return new PagedResponse<AdDto>
        {
            Data = items.Select(a => MapAd(a, baseUrl)).ToList(),
            Meta = new PageMeta
            {
                Page = page, Limit = limit, Total = total,
                TotalPages = totalPages, HasMore = page < totalPages
            }
        };
    }

    // ── Get by id ─────────────────────────────────────────────────────────────
    public async Task<AdDto> GetByIdAsync(string id, string baseUrl, CancellationToken ct)
    {
        var ad = await _ads.GetByIdAsync(id, ct)
            ?? throw ApiException.NotFound("Ad not found.");
        if (!ad.IsActive) throw ApiException.NotFound("Ad not found.");
        return MapAd(ad, baseUrl);
    }

    // ── Create ────────────────────────────────────────────────────────────────
    public async Task<AdDto> CreateAsync(
        string userId, string sellerName,
        CreateAdDto dto, List<IFormFileProxy> photos,
        string baseUrl, CancellationToken ct)
    {
        if (photos.Count == 0)
            throw ApiException.Validation("At least 1 photo is required.");
        if (photos.Count > 5)
            throw ApiException.Validation("Maximum 5 photos allowed.", "too_many_photos");

        ValidateAdFields(dto.Brand, dto.City, dto.Condition);

        var todayCount = await _ads.CountTodayByUserAsync(userId, ct);
        if (todayCount >= 20)
            throw ApiException.TooManyRequests("Daily limit of 20 ads reached.");

        var photoPaths = new List<string>();
        foreach (var photo in photos)
        {
            var saved = await _files.SaveAsync(new FileUploadInput(
                photo.FileName, photo.ContentType, photo.Length, photo.Bytes, "ads", userId), ct);
            photoPaths.Add(saved.Path);
        }

        var condEnum = ParseCondition(dto.Condition);
        var ad = new PhoneAd
        {
            Id = Guid.NewGuid().ToString(),
            Title = dto.Title.Trim(),
            Brand = PhoneConstants.NormaliseBrand(dto.Brand),
            Model = dto.Model.Trim(),
            Price = dto.Price,
            Condition = condEnum,
            City = PhoneConstants.NormaliseCity(dto.City),
            Description = dto.Description.Trim(),
            ContactNumber = dto.ContactNumber,
            PhotoUrlsJson = JsonSerializer.Serialize(photoPaths),
            UserId = userId,
            SellerName = sellerName,
            IsActive = true,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        await _ads.AddAsync(ad, ct);
        return MapAd(ad, baseUrl);
    }

    // ── Update ────────────────────────────────────────────────────────────────
    public async Task<AdDto> UpdateAsync(
        string id, string userId,
        UpdateAdDto dto, List<IFormFileProxy> newPhotos,
        string baseUrl, CancellationToken ct)
    {
        var ad = await _ads.GetByIdAsync(id, ct)
            ?? throw ApiException.NotFound("Ad not found.");
        if (ad.UserId != userId) throw ApiException.Forbidden("You are not allowed to edit this ad.");

        if (dto.Brand is not null) ValidateAdFields(dto.Brand, null, null);
        if (dto.City is not null) ValidateAdFields(null, dto.City, null);
        if (dto.Condition is not null) ValidateAdFields(null, null, dto.Condition);

        if (dto.Title is not null)          ad.Title = dto.Title.Trim();
        if (dto.Brand is not null)          ad.Brand = PhoneConstants.NormaliseBrand(dto.Brand);
        if (dto.Model is not null)          ad.Model = dto.Model.Trim();
        if (dto.Price is not null)          ad.Price = dto.Price.Value;
        if (dto.Condition is not null)      ad.Condition = ParseCondition(dto.Condition);
        if (dto.City is not null)           ad.City = PhoneConstants.NormaliseCity(dto.City);
        if (dto.Description is not null)    ad.Description = dto.Description.Trim();
        if (dto.ContactNumber is not null)  ad.ContactNumber = dto.ContactNumber;

        if (newPhotos.Count > 0)
        {
            if (newPhotos.Count > 5)
                throw ApiException.Validation("Maximum 5 photos allowed.", "too_many_photos");

            // Delete old blobs
            var oldPaths = JsonSerializer.Deserialize<List<string>>(ad.PhotoUrlsJson) ?? [];
            foreach (var p in oldPaths)
                await _files.DeleteByPathAsync(p, ct).ContinueWith(_ => { });

            // Save new blobs
            var newPaths = new List<string>();
            foreach (var photo in newPhotos)
            {
                var saved = await _files.SaveAsync(new FileUploadInput(
                    photo.FileName, photo.ContentType, photo.Length, photo.Bytes, "ads", userId), ct);
                newPaths.Add(saved.Path);
            }
            ad.PhotoUrlsJson = JsonSerializer.Serialize(newPaths);
        }

        ad.UpdatedAt = DateTime.UtcNow;
        await _ads.UpdateAsync(ad, ct);
        return MapAd(ad, baseUrl);
    }

    // ── Soft delete ───────────────────────────────────────────────────────────
    public async Task SoftDeleteAsync(string id, string userId, CancellationToken ct)
    {
        var ad = await _ads.GetByIdAsync(id, ct)
            ?? throw ApiException.NotFound("Ad not found.");
        if (ad.UserId != userId) throw ApiException.Forbidden("You are not allowed to delete this ad.");
        ad.IsActive = false;
        ad.UpdatedAt = DateTime.UtcNow;
        await _ads.UpdateAsync(ad, ct);
    }

    // ── My ads ────────────────────────────────────────────────────────────────
    public async Task<PagedResponse<AdDto>> MyAdsAsync(string userId, string baseUrl, CancellationToken ct)
    {
        var items = await _ads.GetByUserAsync(userId, ct);
        return new PagedResponse<AdDto>
        {
            Data = items.Select(a => MapAd(a, baseUrl)).ToList(),
            Meta = new PageMeta { Total = items.Count }
        };
    }

    // ── Helpers ───────────────────────────────────────────────────────────────
    private static void ValidateAdFields(string? brand, string? city, string? condition)
    {
        if (brand is not null && !PhoneConstants.IsBrandAllowed(brand))
            throw ApiException.Validation("Brand not in allowed list.", "invalid_brand");
        if (city is not null && !PhoneConstants.IsCityAllowed(city))
            throw ApiException.Validation("City not in allowed list.", "invalid_city");
        if (condition is not null && !PhoneConstants.IsConditionAllowed(condition))
            throw ApiException.Validation("Invalid condition value.", "invalid_condition");
    }

    private static ConditionValue ParseCondition(string s) => s switch
    {
        "brandNew" => ConditionValue.BrandNew,
        "likeNew"  => ConditionValue.LikeNew,
        "good"     => ConditionValue.Good,
        "fair"     => ConditionValue.Fair,
        _          => throw ApiException.Validation("Invalid condition value.", "invalid_condition")
    };

    private static string ConditionToString(ConditionValue c) => c switch
    {
        ConditionValue.BrandNew => "brandNew",
        ConditionValue.LikeNew  => "likeNew",
        ConditionValue.Good     => "good",
        ConditionValue.Fair     => "fair",
        _                       => c.ToString()
    };

    private AdDto MapAd(PhoneAd ad, string baseUrl)
    {
        var rawPaths = JsonSerializer.Deserialize<List<string>>(ad.PhotoUrlsJson) ?? [];
        var photoUrls = rawPaths.Select(p => _files.ServeUrl(p, baseUrl)).ToList();

        return new AdDto
        {
            Id = ad.Id,
            Title = ad.Title,
            Brand = ad.Brand,
            Model = ad.Model,
            Price = ad.Price,
            Condition = ConditionToString(ad.Condition),
            City = ad.City,
            Description = ad.Description,
            ContactNumber = ad.ContactNumber,
            PhotoUrls = photoUrls,
            UserId = ad.UserId,
            SellerName = ad.SellerName,
            IsActive = ad.IsActive,
            CreatedAt = ad.CreatedAt.ToString("O"),
            UpdatedAt = ad.UpdatedAt.ToString("O")
        };
    }
}

/// <summary>Abstraction over IFormFile so the service has no ASP.NET dependency.</summary>
public record IFormFileProxy(string FileName, string ContentType, long Length, byte[] Bytes);
